### DISCUSSION TITLE
Special case `Date` instances in `replaceEqualDeep` function

### DISCUSSION THREAD

**Discussion by jcready** [TanStack/query#3559] [Ideas] [answered]
If my query returns a `Date` instance then it means that every time that query gets refetched it will over-write the existing `Date` instance in the cache with a new one even if both `Date` instances are referring to the same point in time. This also means that the cache points to a new reference of the entity (so any hooks that relied on a stable value have to recompute). It would be nice if that could be avoided. https://github.com/tannerlinsley/react-query/blob/07c9ce01236b387d9a5c424b0a76c830a325d4b9/src/core/utils.ts#L340-L373
More generally, it could be beneficial to either handle more data structures (like `Map` and `Set`) and/or allow for the consumer to specify their own `replaceEqualDeep` function somehow.

**Comment by TkDodo** [comment, marked as answer]
you can pass an `isDataEqual` function to `useQuery` or provide it globally. There, you can do your own comparison :)

it's not very well documented, but mentioned here:
https://react-query.tanstack.com/guides/important-defaults

> Structural sharing only works with JSON-compatible values, any other value types will always be considered as changed. If you are seeing performance issues because of large responses for example, you can disable this feature with the config.structuralSharing flag. If you are dealing with non-JSON compatible values in your query responses and still want to detect if data has changed or not, you can define a data compare function with config.isDataEqual.

**Reply by jcready** [reply]
Ah, I should've realized this was already thought of. Thank you!

**Reply by TkDodo** [reply]
we only have this one entry, and then another short mention in the `api` docs, but details are missing there. It should really be in the options list here as well: https://react-query.tanstack.com/reference/useQuery

**Reply by jcready** [reply]
Oh, now I remember why I wanted to be able to set my own `replaceEqualDeep`. It's because we have some non-enumerable properties that get added to the cached object and the current implementation obviously doesn't account for that (I wouldn't expect it to). This means that as soon as a new version of the entity comes across the wire with structural sharing enabled we lose access to that non-enumerable property because `replaceEqualDeep` spreads the cached version.

**Reply by TkDodo** [reply]
in this case, what I would do is read the previous cache entry value in the queryFn and keep the old values that have these non-enumerable properties. Something like:

```
const queryClient = useQueryClient()
const key = ['my', 'key']

useQuery(key, async () => {
  const previousData = queryClient.getQueryData(key)
  const { data } = await axios.get(...)
  if (!previousData) return data
  return mergeAndKeepSomeFieldsFromPreviousData(previousData, data)
})
```
